src/negativity.jl

```diff
@@ -0,0 +1,136 @@
+- `subsys::Int`: an index that indicates which subsystem to compute the negativity for.
+- `logarithmic::Bool`: choose whether to calculate logarithmic negativity or not. Default as `false`
+
+# Returns
+- `N::Real`: The value of negativity.
+
+# Examples
+
+```
+julia> Ψ = 1 / √2 * ( basis(2,0) ⊗ basis(2,0) + basis(2,1) ⊗ basis(2,1) )
+Quantum Object:   type=Ket   dims=[2, 2]   size=(4,)
+4-element Vector{ComplexF64}:
+ 0.7071067811865475 + 0.0im
+                0.0 + 0.0im
+                0.0 + 0.0im
+ 0.7071067811865475 + 0.0im
+
+julia> ρ = ket2dm(Ψ)
+Quantum Object:   type=Operator   dims=[2, 2]   size=(4, 4)   ishermitian=true
+4×4 Matrix{ComplexF64}:
+ 0.5+0.0im  0.0+0.0im  0.0+0.0im  0.5+0.0im
+ 0.0+0.0im  0.0+0.0im  0.0+0.0im  0.0+0.0im
+ 0.0+0.0im  0.0+0.0im  0.0+0.0im  0.0+0.0im
+ 0.5+0.0im  0.0+0.0im  0.0+0.0im  0.5+0.0im
+
+julia> negativity(ρ, 2)
+0.4999999999999998
+```
+"""
+function negativity(ρ::QuantumObject, subsys::Int; logarithmic::Bool=false)
+    mask = fill(false, length(ρ.dims))
+    try
+        mask[subsys] = true
+    catch
+        error("Invalid index of subsys: $subsys")
+    end
+
+    ρ_pt = partial_transpose(ρ, mask)
+    tr_norm = sum(sqrt.(eigvals(ρ_pt' * ρ_pt)))
+
+    if logarithmic
+        return log2(tr_norm)
+    else
+        return (tr_norm - 1) / 2
+    end
+end
+
+@doc raw"""
+    partial_transpose(ρ::QuantumObject, mask::Vector{Bool})
+
+Return the partial transpose of a density matrix ``\rho``, where `mask` is an array/vector with length that equals the length of `ρ.dims`. The elements in `mask` are boolean (`true` or `false`) which indicates whether or not the corresponding subsystem should be transposed.
+
+# Arguments
+- `ρ::QuantumObject`: The density matrix (`ρ.type` must be `OperatorQuantumObject`).
+- `mask::Vector{Bool}`: A boolean vector selects which subsystems should be transposed.
+
+# Returns
+- `ρ_pt::QuantumObject`: The density matrix with the selected subsystems transposed.
+"""
+function partial_transpose(ρ::QuantumObject{T, OperatorQuantumObject}, mask::Vector{Bool}) where T
+    if length(mask) != length(ρ.dims) 
+        error("The length of \`mask\` should be equal to the length of \`ρ.dims\`.")
+    end
+    return _partial_transpose(ρ, mask)
+end
+
+# for dense matrices
+function _partial_transpose(ρ::QuantumObject{<:AbstractMatrix, OperatorQuantumObject}, mask::Vector{Bool})
+    mask2 = [1 + Int(i) for i in mask]
+    # mask2 has elements with values equal to 1 or 2
+    #   1 - the subsystem don't need to be transposed
+    #   2 - the subsystem need be transposed
+
+    nsys = length(mask2)
+    pt_dims = reshape(Vector(1:(2 * nsys)), (nsys, 2))
+    pt_idx  = [
+        [pt_dims[n,     mask2[n]] for n in 1:nsys]; # origin   value in mask2
+        [pt_dims[n, 3 - mask2[n]] for n in 1:nsys]  # opposite value in mask2 (1 -> 2, and 2 -> 1)
+    ]
+    return QuantumObject(
+        reshape(permutedims(reshape(ρ.data, (ρ.dims..., ρ.dims...)), pt_idx), size(ρ)),
```

Comment:
Be aware that I have already implemented the `row_major_reshape` function, if you need it. Because it seems that what you are doing here is very similar.

Reply:
I think it is still a bit different, the tricky part here is that the variable `pt_idx` (i.e., the `perm`utation) contains all the information for partial transposing the basis.
Let's say I have a `QuantumObject` which contains three sub-system (`dims=[N1, N2, N3]`), and I want to partial transpose:
- the first sub-system, then `pt_idx=[4, 2, 3, 1, 5, 6]`
- the second sub-system, then `pt_idx=[1, 5, 3, 4, 2, 6]`
- or even both the first and second sub-system, then `pt_idx=[4, 5, 3, 1, 2, 6]`

I looked into the definition of `row_major_reshape`:
```
PermutedDimsArray(reshape(Q, reverse(shapes)...), (length(shapes):-1:1))
```
The argument `perm` in `PermutedDimsArray` is fixed to be `length(shapes):-1:1`, where `shapes` here equals to `(dims..., dims...)=(N1, N2, N3, N1, N2, N3)`. So I think `row_major_reshape` doesn't fit the case I need here ?

However, I didn't know `PermutedDimsArray(A, perm) -> B` in the beginning. The official documentation says that it is similar to `permutedims`, except that no copying occurs (`B` shares storage with `A`). Therefore, I think I should use `PermutedDimsArray` instead of `permutedims`.

Reply:
Yes, I think that it is better to use `PermutedDimsArray`

Reply:
updated in comment cd7d7a79d0fca3cbf6ef9c85877066be09f960d4